woocommerce/woocommerce-ios
# IPP Analytics: track Interac events

Closes: #6655 


### Description


Please note that there is a p2 post pdfdoF-ET-p2 with next steps that might change the collect payment flow updated in this PR.

This PR tracks 3 new events for Moose on the Loose for Interac payment method:

Event | Properties | Trigger
-- | -- | --
*_card_interac_collect_payment_success | | Tracked right after processing succeeds.
*_interac_refund_success | | Tracked when an Interac in-person refund succeeds on the client side.
*_interac_refund_failed | error | Tracked when an Interac in-person refund fails on the client side.

The trickiest part is to track `*_card_interac_collect_payment_success`, which is logged right after client-side processing. However, the client-side processing code is currently in a chain of steps:

https://github.com/woocommerce/woocommerce-ios/blob/d5586d2cf086852b4438fed953f9aaa0429dddc4/Hardware/Hardware/CardReader/StripeCardReader/StripeCardReaderService.swift#L236-L247

After client-side processing, there is still one step to wait for chip card removal before server-side processing as in the full collect payment flow below.

<img width="1401" alt="ipp-collect-payment-detailed-background" src="https://user-images.githubusercontent.com/1945542/163944565-191809ad-56a9-42a0-93ad-6597ea2b1a6c.png">

In order to track `*_card_interac_collect_payment_success` right after client-side processing, I added another publisher `processingCompleted: AnyPublisher<PaymentIntent, Never>` to return in addition to the original `future`. If the proposed iOS changes in pdfdoF-ET-p2 are approved, then we are probably going to return the `future` after client-side processing and then wait for chip card removal and dispatch server-side processing in parallel.

### Testing instructions


Prerequisites: the store is in Canada and eligible for IPP (setup instructions PdfdoF-D-p2)

- Go to the orders tab
- Create an order or a simple payment
- Go to order details
- Tap “Collect Payment” CTA
- Connect to a WisePad 3 card reader
- Insert the Interac test card and enter the pin code --> after the "Remove card" prompt appears, you should see `🔵 Tracked card_interac_collect_payment_success, properties: [AnyHashable("country"): "CA", AnyHashable("card_reader_model"): "WISEPAD_3", AnyHashable("plugin_slug"): "woocommerce-payments", ...]` in the console
- Remove the Interac test card
- Back in order details, tap "Issue Refund" CTA
- Try inserting a non-Interac card to generate an error --> after the "Remove card" prompt appears, you should see `🔵 Tracked interac_refund_failed, properties: [AnyHashable("country"): "CA", AnyHashable("card_reader_model"): "WISEPAD_3", AnyHashable("plugin_slug"): "woocommerce-payments", ...]` in the console
- Remove the card in the previous step
- Tap "Try again" to retry refund
- Insert the Interac test card and enter the pin code --> after the "Remove card" prompt appears, you should see `🔵 Tracked interac_refund_success, properties: [AnyHashable("country"): "CA", AnyHashable("card_reader_model"): "WISEPAD_3", AnyHashable("plugin_slug"): "woocommerce-payments", ...]` in the console


---
